### PR TITLE
Add a setting to configure centered layout fixed width

### LINKED ISSUE #54731: Give the option to fix centered layout width.
Previously the centered layout in zen mode had a fixed width, meaning that if you resize the window the centered section will remain the same as long as it still fits in the resized window. The behaviour now is for the centered section width to be proportional to the size of the window. See below:

image

On resize, old behaviour/new behaviour:

image

### PR DESCRIPTION
**Pull request by FancyBanana**
This PR fixes #54731. 
This is the same code as in PR #104942, but updated to allow setting fixed layout as option. The default behavior remains the same.

This adds a new configuration option `workbench.editor.centeredLayoutFixedWidth` that allows to change the way centered layout reacts to window width change. The default is to stretch/squeeze editor in proportion to window. The new `fixed` behavior attempts to preserve the width. 

The layout width is preserved between the sessions just like with dynamic width; and when switching between the two the layout size is recalculated to preserve the current width.

### CODE REVIEW

**Comment by FancyBanana** [comment]
@microsoft-github-policy-service agree

**Comment by FancyBanana** [comment]
Is there any change of this being merged before the next update?

**Comment by sbatten** [comment]
@FancyBanana not before the next update as we are in testing phase right now. But it could make it into the next one

**Comment by bpasero** [comment]
Marking as draft since I spotted a `TODO` in the code.

**Comment by FancyBanana** [comment]
@bpasero This is not as much "I will adress this later" as "I did my best but maybe vscode folks can offer something better"

**Comment by bpasero** [comment]
I tested centered layout with the setting enabled and disabled and could not find any difference? What do I need to look at?

File: src/vs/workbench/browser/parts/editor/editorPart.ts [hunk 1]
```diff
@@ -172,6 +172,8 @@ export class EditorPart extends Part implements IEditorGroupsService, IEditorGro
 
 		this._partOptions = newPartOptions;
 
+		this.centeredLayoutWidget.setFixedWidth(this._partOptions.centeredLayoutFixedWidth ?? false);
```
Comment (bpasero) [inline src/vs/workbench/browser/parts/editor/editorPart.ts:175]: I think a little bit nicer would be to listen to the event from here:

https://github.com/microsoft/vscode/blob/4f6f7d50b9941ab4dcd6280f49584c1c66e7b17a/src/vs/workbench/browser/parts/editor/editorPart.ts#L859

And forward into the widget.

**Review by bpasero — CHANGES_REQUESTED** [review]

**Comment by FancyBanana** [comment]
@bpasero  Here's a demonstration of what this is supposed to do: 

https://user-images.githubusercontent.com/7935263/206711758-a08eda3c-1a90-4f4d-85ab-54811ff77cc8.mp4

File: src/vs/workbench/browser/parts/editor/editorPart.ts [hunk 2]
```diff
@@ -855,6 +855,7 @@ export class EditorPart extends Part implements IEditorGroupsService, IEditorGro
 
 		// Centered layout widget
 		this.centeredLayoutWidget = this._register(new CenteredViewLayout(this.container, this.gridWidgetView, this.profileMemento[EditorPart.EDITOR_PART_CENTERED_VIEW_STORAGE_KEY]));
+		this.onDidChangeEditorPartOptions(e => this.centeredLayoutWidget.setFixedWidth(e.newPartOptions.centeredLayoutFixedWidth ?? false));
```
Comment (bpasero) [inline src/vs/workbench/browser/parts/editor/editorPart.ts:858]: This misses a `this._register` to cleanup on displse.

**Review by bpasero — DISMISSED** [review]

**Review by bpasero — COMMENTED** [review]
I still do not see the difference, but I am also not a big user of this feature, so leaving it up to SteVen.

I only reviewed the changes in `editorPart.ts` and think they are fine.

**Comment by bpasero** [comment]
I defer the main review to Steven because most changes are in centered layout area.

**Comment by FancyBanana** [comment]
@sbatten Hello :)

**Comment by FancyBanana** [comment]
Is this PR still considered for merge? It's been almost a month since it was commented on.

**Review by sbatten — APPROVED** [review]

File: src/vs/base/browser/ui/centered/centeredViewLayout.ts [hunk 3]
```diff
@@ -91,21 +103,51 @@ export class CenteredViewLayout implements IDisposable {
...
 		}
-		this.splitView.resizeView(0, this.state.leftMarginRatio * this.width);
-		this.splitView.resizeView(2, this.state.rightMarginRatio * this.width);
+		if (this.centeredLayoutFixedWidth) {
+
+			const centerViewWidth = Math.min(this.width, this.state.targetWidth);
+			const marginWidthFloat = (this.width - centerViewWidth) / 2;
+			this.splitView.resizeView(0, Math.floor(marginWidthFloat));
+			this.splitView.resizeView(1, centerViewWidth);
+			this.splitView.resizeView(2, Math.ceil(marginWidthFloat));
```
Comment (sbatten) [inline src/vs/base/browser/ui/centered/centeredViewLayout.ts:126]: @joaomoreno accepting this PR, but I think there is something better we can do with split views here. I feel like we should be able to set LayoutPriority based on the setting and the split view should handle this, but LayoutPriority is not honored in proportional layouting. With centered view fixed width, I would expect the margins to have a high layout priority and the centered view priority to be low and that would be all that's needed.

**Review by joyceerhl — APPROVED** [review]

File: src/vs/base/browser/ui/centered/centeredViewLayout.ts (hunk 3 above)
Comment (joaomoreno) [reply src/vs/base/browser/ui/centered/centeredViewLayout.ts:126]: I agree. This should've been contributed at the splitview level in some other shape or form. It seems that when this setting is `true`, multiple layouts will happen in a row now, right?

**Pull request merged**